File: Example/FWFramework/Classes/Module/Test/Toolkit/TestKeyboardController.m
//
//  TestKeyboardController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestKeyboardController.h"
#import "AppSwift.h"
@import FWFramework;

@interface TestKeyboardController () <FWScrollViewController, UITextFieldDelegate, UITextViewDelegate>

FWPropertyStrong(UITextField *, mobileField);

FWPropertyStrong(UITextField *, passwordField);

FWPropertyStrong(UITextView *, textView);

FWPropertyStrong(UITextView *, inputView);

FWPropertyStrong(UIButton *, submitButton);

FWPropertyStrong(FWPopupMenu *, popupMenu);

FWPropertyAssign(BOOL, canScroll);
FWPropertyAssign(BOOL, dismissOnDrag);
FWPropertyAssign(BOOL, useScrollView);
FWPropertyCopy(NSString *, appendString);

@end

@implementation TestKeyboardController

- (void)setCanScroll:(BOOL)canScroll
{
    _canScroll = canScroll;
    [self.view endEditing:YES];
    [self renderData];
}

- (void)setDismissOnDrag:(BOOL)dismissOnDrag
{
    _dismissOnDrag = dismissOnDrag;
    [self.view endEditing:YES];
    self.scrollView.keyboardDismissMode = UIScrollViewKeyboardDismissModeOnDrag;
}

- (void)setUseScrollView:(BOOL)useScrollView
{
    _useScrollView = useScrollView;
    [self.view endEditing:YES];
    self.navigationItem.title = useScrollView ? @"UIScrollView+FWKeyboard" : @"UITextField+FWKeyboard";
    NSArray<UIView *> *textInputs = @[self.mobileField, self.passwordField, self.textView, self.inputView];
    for (UIView *textInput in textInputs) {
        ((UITextField *)textInput).fw_keyboardScrollView = useScrollView ? self.scrollView : nil;
    }
}

- (void)setupSubviews
{
    self.scrollView.backgroundColor = [AppTheme tableColor];
    
    UITextField *textFieldAppearance = [UITextField appearanceWhenContainedInInstancesOfClasses:@[[TestKeyboardController class]]];
    UITextView *textViewAppearance = [UITextView appearanceWhenContainedInInstancesOfClasses:@[[TestKeyboardController class]]];
    textFieldAppearance.fw_keyboardManager = YES;
    textFieldAppearance.fw_touchResign = YES;
    textFieldAppearance.fw_keyboardResign = YES;
    textFieldAppearance.fw_reboundDistance = 200;
    textViewAppearance.fw_keyboardManager = YES;
    textViewAppearance.fw_touchResign = YES;
    textViewAppearance.fw_keyboardResign = YES;
    textViewAppearance.fw_reboundDistance = 200;
    
    UITextField *mobileField = [self createTextField];
    mobileField.tag = 1;
    self.mobileField = mobileField;
    mobileField.fw_maxUnicodeLength = 10;
    mobileField.fw_menuDisabled = YES;
    mobileField.placeholder = @"禁止粘贴，最多10个中文";
    mobileField.keyboardType = UIKeyboardTypeDefault;
    mobileField.returnKeyType = UIReturnKeyNext;
    [self.contentView addSubview:mobileField];
    [mobileField fw_pinEdgeToSuperview:NSLayoutAttributeLeft withInset:15];
    [mobileField fw_pinEdgeToSuperview:NSLayoutAttributeRight withInset:15];
    [mobileField fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    
    UITextField *passwordField = [self createTextField];
    passwordField.tag = 2;
    self.passwordField = passwordField;
    passwordField.delegate = self;
    passwordField.fw_maxLength = 20;
    passwordField.placeholder = @"仅数字和字母转大写，最多20个英文";
    passwordField.keyboardType = UIKeyboardTypeDefault;
    passwordField.returnKeyType = UIReturnKeyNext;
    mobileField.fw_returnNext = YES;
    FWWeakifySelf();
    mobileField.fw_nextResponder = ^(UITextField *textField){
        FWStrongifySelf();
        return self.passwordField;
    };
    [mobileField fw_addToolbarWithTitle:[NSAttributedString fw_attributedString:mobileField.placeholder withFont:[UIFont systemFontOfSize:13.0]] doneBlock:nil];
    [self.contentView addSubview:passwordField];
    [passwordField fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:mobileField];
    [passwordField fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    
    UITextView *textView = [self createTextView];
    textView.tag = 3;
    self.textView = textView;
    textView.delegate = self;
    textView.backgroundColor = [AppTheme backgroundColor];
    textView.fw_maxUnicodeLength = 10;
    textView.fw_placeholder = @"问题，最多10个中文";
    textView.returnKeyType = UIReturnKeyNext;
    passwordField.fw_returnNext = YES;
    passwordField.fw_previousResponderTag = 1;
    passwordField.fw_nextResponderTag = 3;
    [passwordField fw_addToolbarWithTitle:[NSAttributedString fw_attributedString:passwordField.placeholder withFont:[UIFont systemFontOfSize:13.0]] doneBlock:nil];
    [self.contentView addSubview:textView];
    [textView fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:passwordField withOffset:15];
    [textView fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    
    UITextView *inputView = [self createTextView];
    inputView.tag = 4;
    self.inputView = inputView;
    inputView.backgroundColor = [AppTheme backgroundColor];
    inputView.fw_maxLength = 20;
    inputView.fw_menuDisabled = YES;
    inputView.fw_placeholder = @"建议，最多20个英文";
    inputView.returnKeyType = UIReturnKeyDone;
    inputView.fw_returnResign = YES;
    inputView.fw_keyboardDistance = 80;
    textView.fw_returnNext = YES;
    textView.fw_previousResponderTag = 2;
    textView.fw_nextResponderTag = 4;
    [textView fw_addToolbarWithTitle:[NSAttributedString fw_attributedString:textView.fw_placeholder withFont:[UIFont systemFontOfSize:13.0]] doneBlock:nil];
    inputView.fw_delegate = self;
    inputView.fw_previousResponderTag = 3;
    [inputView fw_addToolbarWithTitle:[NSAttributedString fw_attributedString:inputView.fw_placeholder withFont:[UIFont systemFontOfSize:13.0]] doneBlock:nil];
    [self.contentView addSubview:inputView];
    [inputView fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:textView withOffset:15];
    [inputView fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    
    UIButton *submitButton = [AppTheme largeButton];
    self.submitButton = submitButton;
    [submitButton setTitle:@"提交" forState:UIControlStateNormal];
    [submitButton fw_addTouchTarget:self action:@selector(onSubmit)];
    [self.contentView addSubview:submitButton];
    [submitButton fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:inputView withOffset:15];
    [submitButton fw_pinEdgeToSuperview:NSLayoutAttributeBottom withInset:15];
    [submitButton fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
}

- (void)setupLayout
{
    FWWeakifySelf();
    self.mobileField.fw_autoCompleteBlock = ^(NSString * _Nonnull text) {
        FWStrongifySelf();
        if (text.length < 1) {
            [self.popupMenu dismiss];
        } else {
            [self.popupMenu dismiss];
            self.popupMenu = [FWPopupMenu showRelyOnView:self.mobileField
                                                  titles:@[text]
                                                   icons:nil
                                               menuWidth:self.mobileField.fw_width
                                           otherSettings:^(FWPopupMenu * _Nonnull popupMenu) {
                popupMenu.showMaskView = NO;
            }];
        }
    };
    
    self.inputView.fw_autoCompleteBlock = ^(NSString * _Nonnull text) {
        FWStrongifySelf();
        if (text.length < 1) {
            [self.popupMenu dismiss];
        } else {
            [self.popupMenu dismiss];
            self.popupMenu = [FWPopupMenu showRelyOnView:self.inputView
                                                  titles:@[text]
                                                   icons:nil
                                               menuWidth:self.inputView.fw_width
                                           otherSettings:^(FWPopupMenu * _Nonnull popupMenu) {
                popupMenu.showMaskView = NO;
            }];
        }
    };
    
    [self fw_setRightBarItem:@"切换" block:^(id sender) {
        FWStrongifySelf();
        [self fw_showSheetWithTitle:nil message:nil cancel:@"取消" actions:@[@"切换滚动", @"切换滚动时收起键盘", @"切换滚动视图", @"自动添加-"] actionBlock:^(NSInteger index) {
            FWStrongifySelf();
            if (index == 0) {
                self.canScroll = !self.canScroll;
            } else if (index == 1) {
                self.dismissOnDrag = !self.dismissOnDrag;
            } else if (index == 2) {
                self.useScrollView = !self.useScrollView;
            } else {
                self.appendString = self.appendString ? nil : @"-";
            }
        }];
    }];
    
    [self renderData];
}

- (void)renderData
{
    CGFloat marginTop = FWScreenHeight - (390 + 15 + FWTopBarHeight + UIScreen.fw_safeAreaInsets.bottom);
    CGFloat topInset = self.canScroll ? FWScreenHeight : marginTop;
    [self.mobileField fw_pinEdgeToSuperview:NSLayoutAttributeTop withInset:topInset];
}

- (UITextView *)createTextView
{
    UITextView *textView = [UITextView new];
    textView.font = [UIFont fw_fontOfSize:15];
    textView.textColor = [AppTheme textColor];
    textView.tintColor = AppTheme.textColor;
    textView.fw_cursorRect = CGRectMake(0, 0, 2, 0);
    [textView fw_setBorderColor:[AppTheme borderColor] width:0.5 cornerRadius:5];
    [textView fw_setDimension:NSLayoutAttributeWidth toSize:FWScreenWidth - 15 * 2];
    [textView fw_setDimension:NSLayoutAttributeHeight toSize:100];
    return textView;
}

- (UITextField *)createTextField
{
    UITextField *textField = [UITextField new];
    textField.font = [UIFont fw_fontOfSize:15];
    textField.textColor = [AppTheme textColor];
    textField.tintColor = AppTheme.textColor;
    textField.fw_cursorRect = CGRectMake(0, 0, 2, 0);
    textField.clearButtonMode = UITextFieldViewModeWhileEditing;
    [textField fw_setBorderView:UIRectEdgeBottom color:[AppTheme borderColor] width:0.5];
    [textField fw_setDimension:NSLayoutAttributeWidth toSize:FWScreenWidth - 15 * 2];
    [textField fw_setDimension:NSLayoutAttributeHeight toSize:50];
    return textField;
}

#pragma mark - UITextFieldDelegate

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
    if ([string isEqualToString:@"\n"]) {
        return YES;
    }
    
    NSString *allowedChars = @"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
    NSCharacterSet *characterSet = [[NSCharacterSet characterSetWithCharactersInString:allowedChars] invertedSet];
    NSString *filterString = [[string componentsSeparatedByCharactersInSet:characterSet] componentsJoinedByString:@""];
    if (![string isEqualToString:filterString]) {
        return NO;
    }
    
    if (string.length > 0) {
        NSString *replaceString = string.uppercaseString;
        if (self.appendString) replaceString = [replaceString stringByAppendingString:self.appendString];
        NSString *filterText = [textField.text stringByReplacingCharactersInRange:range withString:replaceString];
        textField.text = [textField fw_filterText:filterText];
        
        NSInteger offset = range.location + replaceString.length;
        if (offset > textField.fw_maxLength) offset = textField.fw_maxLength;
        [textField fw_moveCursor:offset];
        return NO;
    }
    
    return YES;
}

#pragma mark - Action

- (void)onSubmit
{
    NSLog(@"点击了提交");
}

@end

File: Example/FWFramework/Classes/Module/Test/Toolkit/TestKeyboardController.swift
//
//  TestKeyboardController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/28.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestKeyboardController: UIViewController, ScrollViewControllerProtocol, UITextFieldDelegate, UITextViewDelegate {
    
    private var canScroll = false {
        didSet {
            view.endEditing(true)
            renderData()
        }
    }
    
    private var dismissOnDrag = false {
        didSet {
            view.endEditing(true)
            scrollView.keyboardDismissMode = dismissOnDrag ? .onDrag : .none
        }
    }
    
    private var useScrollView = false {
        didSet {
            view.endEditing(true)
            navigationItem.title = useScrollView ? "UIScrollView+FWKeyboard" : "UITextField+FWKeyboard"
            mobileField.app.keyboardScrollView = useScrollView ? scrollView : nil
            passwordField.app.keyboardScrollView = useScrollView ? scrollView : nil
            textView.app.keyboardScrollView = useScrollView ? scrollView : nil
            descView.app.keyboardScrollView = useScrollView ? scrollView : nil
        }
    }
    
    private var appendString = ""
    private var popupMenu: PopupMenu?
    
    private lazy var mobileField: UITextField = {
        let result = createTextField()
        result.tag = 1
        result.app.maxUnicodeLength = 10
        result.app.menuDisabled = true
        result.placeholder = "禁止粘贴，最多10个中文"
        result.keyboardType = .default
        result.returnKeyType = .next
        result.textContentType = .telephoneNumber
        return result
    }()
    
    private lazy var passwordField: UITextField = {
        let result = createTextField()
        result.tag = 2
        result.delegate = self
        result.app.maxLength = 20
        result.placeholder = "仅数字和字母转大写，最多20个英文"
        result.keyboardType = .default
        result.returnKeyType = .next
        result.textContentType = .password
        return result
    }()
    
    private lazy var textView: UITextView = {
        let result = createTextView()
        result.tag = 3
        result.backgroundColor = AppTheme.backgroundColor
        result.app.maxLength = 200
        result.app.placeholder = "问题\n最多200个字符"
        result.app.lineHeight = 25
        result.app.textChangedBlock = { [weak self] text in
            self?.countLabel.text = "\(self?.textView.app.actualNumberOfLines ?? 0)行 \(text.count)/\(self?.textView.app.maxLength ?? 0)字"
        }
        // result.returnKeyType = .next
        return result
    }()
    
    private lazy var countLabel: UILabel = {
        let result = UILabel()
        result.font = UIFont.app.font(ofSize: 13)
        result.textColor = AppTheme.textColor
        result.textAlignment = .right
        result.text = "0行 0/\(textView.app.maxLength)字"
        return result
    }()
    
    private lazy var descView: UITextView = {
        let result = createTextView()
        result.tag = 4
        result.backgroundColor = AppTheme.backgroundColor
        result.app.maxLength = 20
        result.app.menuDisabled = true
        result.app.placeholder = "仅数字和字母转大写，最多20个英文"
        result.returnKeyType = .done
        result.app.returnResign = true
        result.app.keyboardDistance = 80
        result.app.delegate = self
        return result
    }()
    
    private lazy var submitButton: UIButton = {
        let result = AppTheme.largeButton()
        result.setTitle("提交", for: .normal)
        result.app.addTouch(target: self, action: #selector(onSubmit))
        return result
    }()
    
    func setupSubviews() {
        scrollView.backgroundColor = AppTheme.tableColor
        
        let textFieldAppearance = UITextField.appearance(whenContainedInInstancesOf: [TestKeyboardController.self])
        let textViewAppearance = UITextView.appearance(whenContainedInInstancesOf: [TestKeyboardController.self])
        textFieldAppearance.app.keyboardManager = true
        textFieldAppearance.app.touchResign = true
        textFieldAppearance.app.keyboardResign = true
        textFieldAppearance.app.reboundDistance = 200
        textViewAppearance.app.keyboardManager = true
        textViewAppearance.app.touchResign = true
        textViewAppearance.app.keyboardResign = true
        textViewAppearance.app.reboundDistance = 200
        
        contentView.addSubview(mobileField)
        mobileField.app.layoutChain
            .left(15)
            .right(15)
            .centerX()
        mobileField.app.returnNext = true
        mobileField.app.nextResponder = { [weak self] textField in
            return self?.passwordField
        }
        mobileField.app.addToolbar(title: NSAttributedString.app.attributedString(mobileField.placeholder ?? "", font: UIFont.systemFont(ofSize: 13)), doneBlock: nil)
        
        contentView.addSubview(passwordField)
        passwordField.app.layoutChain
            .centerX()
            .top(toViewBottom: mobileField)
        passwordField.app.returnNext = true
        passwordField.app.previousResponderTag = 1
        passwordField.app.nextResponderTag = 3
        passwordField.app.addToolbar(title: NSAttributedString.app.attributedString(passwordField.placeholder ?? "", font: UIFont.systemFont(ofSize: 13)), doneBlock: nil)
        
        contentView.addSubview(textView)
        textView.app.layoutChain
            .centerX()
            .top(toViewBottom: passwordField, offset: 15)
        // textView.app.returnNext = true
        textView.app.previousResponderTag = 2
        textView.app.nextResponderTag = 4
        textView.app.addToolbar(title: NSAttributedString.app.attributedString(textView.app.placeholder ?? "", font: UIFont.systemFont(ofSize: 13)), doneBlock: nil)
        
        contentView.addSubview(countLabel)
        countLabel.app.layoutChain
            .top(toViewBottom: textView, offset: 15)
            .right(15)
        
        contentView.addSubview(descView)
        descView.app.previousResponderTag = 3
        descView.app.addToolbar(title: NSAttributedString.app.attributedString(descView.app.placeholder ?? "", font: UIFont.systemFont(ofSize: 13)), doneBlock: nil)
        descView.app.layoutChain
            .centerX()
            .top(toViewBottom: countLabel, offset: 15)
        
        contentView.addSubview(submitButton)
        submitButton.app.layoutChain
            .centerX()
            .bottom(15)
            .top(toViewBottom: descView, offset: 15)
    }
    
    func setupLayout() {
        mobileField.app.autoCompleteBlock = { [weak self] text in
            guard let self = self else { return }
            if text.isEmpty {
                self.popupMenu?.dismiss()
            } else {
                self.popupMenu?.dismiss()
                self.popupMenu = PopupMenu.show(relyOn: self.mobileField, titles: [text], icons: nil, menuWidth: self.mobileField.app.width, customize: { popupMenu in
                    popupMenu.maskViewColor = .clear
                })
            }
        }
        
        descView.app.autoCompleteBlock = { [weak self] text in
            guard let self = self else { return }
            if text.isEmpty {
                self.popupMenu?.dismiss()
            } else {
                self.popupMenu?.dismiss()
                self.popupMenu = PopupMenu.show(relyOn: self.descView, titles: [text], icons: nil, menuWidth: self.descView.app.width, customize: { popupMenu in
                    popupMenu.maskViewColor = .clear
                })
            }
        }
        
        app.setRightBarItem("切换") { [weak self] _ in
            self?.app.showSheet(title: nil, message: nil, cancel: "取消", actions: ["切换滚动", "切换滚动时收起键盘", "切换滚动视图", "自动添加-"], currentIndex: -1, actionBlock: { index in
                guard let self = self else { return }
                if index == 0 {
                    self.canScroll = !self.canScroll
                } else if index == 1 {
                    self.dismissOnDrag = !self.dismissOnDrag
                } else if index == 2 {
                    self.useScrollView = !self.useScrollView
                } else {
                    self.appendString = !self.appendString.isEmpty ? "" : "-"
                }
            })
        }
        
        renderData()
    }
    
    private func renderData() {
        let marginTop = APP.screenHeight - (390 + 15 + APP.topBarHeight + UIScreen.app.safeAreaInsets.bottom)
        let topInset = canScroll ? APP.screenHeight : marginTop
        mobileField.app.pinEdge(toSuperview: .top, inset: topInset)
    }
    
    private func createTextField() -> UITextField {
        let result = UITextField()
        result.font = UIFont.app.font(ofSize: 15)
        result.textColor = AppTheme.textColor
        result.tintColor = AppTheme.textColor
        result.app.cursorRect = CGRect(x: 0, y: 0, width: 2, height: 0)
        result.clearButtonMode = .whileEditing
        result.app.setBorderView(.bottom, color: AppTheme.borderColor, width: 0.5)
        result.app.setDimension(.width, size: APP.screenWidth - 30)
        result.app.setDimension(.height, size: 50)
        return result
    }
    
    private func createTextView() -> UITextView {
        let result = UITextView()
        result.font = UIFont.app.font(ofSize: 15)
        result.textColor = AppTheme.textColor
        result.tintColor = AppTheme.textColor
        result.app.cursorRect = CGRect(x: 0, y: 0, width: 2, height: 0)
        result.app.setBorderColor(AppTheme.borderColor, width: 0.5, cornerRadius: 5)
        result.app.setDimension(.width, size: APP.screenWidth - 30)
        result.app.setDimension(.height, size: 100)
        return result
    }
    
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        if string == "\n" {
            return true
        }
        
        let allowedChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
        let characterSet = CharacterSet(charactersIn: allowedChars).inverted
        let filterString = string.components(separatedBy: characterSet).joined(separator: "")
        if string != filterString {
            return false
        }
        
        if !string.isEmpty {
            var replaceString = string.uppercased()
            if !appendString.isEmpty {
                replaceString = replaceString.appending(appendString)
            }
            let curText = (textField.text ?? "") as NSString
            let filterText = curText.replacingCharacters(in: range, with: replaceString)
            textField.text = textField.app.filterText(filterText)
            
            var offset = range.location + replaceString.count
            if offset > textField.app.maxLength {
                offset = textField.app.maxLength
            }
            textField.app.moveCursor(offset)
            return false
        }
        
        return true
    }
    
    func textViewShouldBeginEditing(_ textView: UITextView) -> Bool {
        print("textViewShouldBeginEditing")
        return true
    }
    
    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText string: String) -> Bool {
        if string == "\n" {
            return true
        }
        
        let allowedChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
        let characterSet = CharacterSet(charactersIn: allowedChars).inverted
        let filterString = string.components(separatedBy: characterSet).joined(separator: "")
        if string != filterString {
            return false
        }
        
        if !string.isEmpty {
            var replaceString = string.uppercased()
            if !appendString.isEmpty {
                replaceString = replaceString.appending(appendString)
            }
            let curText = (textView.text ?? "") as NSString
            let filterText = curText.replacingCharacters(in: range, with: replaceString)
            textView.text = textView.app.filterText(filterText)
            
            var offset = range.location + replaceString.count
            if offset > textView.app.maxLength {
                offset = textView.app.maxLength
            }
            textView.app.moveCursor(offset)
            return false
        }
        
        return true
    }
    
    @objc func onSubmit() {
        view.endEditing(true)
        app.showMessage(text: "点击了提交")
    }
    
}
